src/disasm.rs

```diff
@@ -0,0 +1,524 @@
+                    [[:space:]]
+                    (?P<opcode>  [[[:lower:]][[:digit:]][[:space:]]]*)
+                    (?P<operands>[[:space:]]+[^[[:space:]]]+)??
+                    ([[:space:]]+\#(?P<comment>.*))??
+                    [[:space:]]*
+                    $"
+            )
+            .unwrap();
+        }
+
+        if let Some(sec_name) = RE_SECTION
+            .captures(&line)
+            .and_then(|cap| cap.name("sec_name").map(|sec| sec.as_str()))
+        {
+            self.add_section(Section::new(sec_name));
+            Ok(())
+        } else if let Some(sym_name) = RE_SYMBOL
+            .captures(&line)
+            .and_then(|cap| cap.name("sym_name").map(|sym| sym.as_str()))
+        {
+            self.add_symbol(Symbol::new(sym_name))
+        } else if let Some(ins_cap) = RE_INSTRUCTION.captures(&line) {
+            let opcode = ins_cap.name("opcode").map_or("", |m| m.as_str()).trim();
+            let operands = ins_cap.name("operands").map_or("", |m| m.as_str()).trim();
+            let comment = ins_cap.name("comment").map_or("", |m| m.as_str()).trim();
+            self.add_instruction(Instruction::new(opcode, operands, comment))
+        } else {
+            Err(format!(
+                "Unrecognized format for the following line: '{line}'"
+            ))
+        }
+    }
+
+    fn add_section(&mut self, section: Section) {
+        self.sections.push(section);
+    }
+
+    fn add_symbol(&mut self, symbol: Symbol) -> Result<(), String> {
+        self.sections
+            .last_mut()
+            .ok_or("Attempted to add a symbol without first defining a section")?
+            .add_symbol(symbol);
+        Ok(())
+    }
+
+    fn add_instruction(&mut self, instruction: Instruction) -> Result<(), String> {
+        self.sections
+            .last_mut()
+            .ok_or("Attempted to add an instruction without first defining a section")?
+            .add_instruction(instruction)
+    }
+
+    fn sort_sections(&mut self) {
+        for section in &mut self.sections {
+            section.sort_symbols();
+        }
+        self.sections.sort_by(|a, b| a.get_name().cmp(b.get_name()));
+    }
+}
+
+impl fmt::Display for Disasm {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let joined = self
+            .sections
+            .iter()
+            .map(|sec| sec.to_string())
+            .collect::<Vec<_>>()
+            .join("");
+        write!(f, "{}", joined)
+    }
+}
+
+impl TryFrom<BufReader<File>> for Disasm {
+    type Error = String;
+
+    fn try_from(buffer: BufReader<File>) -> Result<Self, Self::Error> {
+        let lines = buffer
+            .lines()
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|msg| format!("Error reading a line of the disassembly file :{msg}"))?;
+        Disasm::from_lines(lines)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn from_lines_empty_vector() {
```

Comment:
State clearly in the test name if it is supposed to fail or succeed, e.g. `empty_vector_from_lines_fails` in this case